wrappers/services.go

```diff
@@ -288,3 +342,61 @@ WantedBy={{.ServicesTarget}}
 
 	return templateOut.Bytes()
 }
+
+func genSocketFile(appInfo *snap.AppInfo, socketName string) []byte {
+	socketTemplate := `[Unit]
+# Auto-generated, DO NO EDIT
+Description=Socket {{.SocketName}} for snap application {{.App.Snap.Name}}.{{.App.Name}}
+Requires={{.MountUnit}}
+Wants={{.PrerequisiteTarget}}
+After={{.MountUnit}} {{.PrerequisiteTarget}}
+X-Snappy=yes
+
+[Socket]
+Service={{.ServiceFileName}}
+ListenStream={{.SocketInfo.ListenStream}}
+{{if .SocketInfo.SocketMode}}SocketMode={{.SocketInfo.SocketMode | printf "%04o"}}{{end}}
+
+[Install]
+WantedBy={{.SocketsTarget}}
+`
+	var templateOut bytes.Buffer
+	t := template.Must(template.New("socket-wrapper").Parse(socketTemplate))
+
+	wrapperData := struct {
+		App                *snap.AppInfo
+		ServiceFileName    string
+		PrerequisiteTarget string
+		SocketsTarget      string
+		MountUnit          string
+		SocketName         string
+		SocketInfo         *snap.SocketInfo
+	}{
+		App:                appInfo,
+		ServiceFileName:    filepath.Base(appInfo.ServiceFile()),
+		SocketsTarget:      systemd.SocketsTarget,
+		PrerequisiteTarget: systemd.PrerequisiteTarget,
+		MountUnit:          filepath.Base(systemd.MountUnitPath(appInfo.Snap.MountDir())),
+		SocketName:         socketName,
+		SocketInfo:         appInfo.Sockets[socketName],
+	}
+
+	if err := t.Execute(&templateOut, wrapperData); err != nil {
+		// this can never happen, except we forget a variable
+		logger.Panicf("Unable to execute template: %v", err)
+	}
+
+	return templateOut.Bytes()
+}
+
+func generateSnapSocketFiles(app *snap.AppInfo) (*map[string][]byte, error) {
+	if err := snap.ValidateApp(app); err != nil {
```

Comment:
Way too late for validation. This should be validated right at the entrance, in the same place we validate everything else about those structures. We should never have in memory an invalid value.

Reply:
I followed the pattern of generateSnapServiceFile, which also calls `ValidateApp`

Reply:
I wonder if I should pull the `Validate()` call out of `generateSnapServiceFile()` and put it directly in `AddSnapServices()`

Reply:
Okay, it's fine to leave it that way then as I may be missing something, but in theory this should be unnecessary in both cases as the app instance here should definitely already have been validated if it is in memory and reached this stage. Ideally we never want instances walking around in the process that are invalid.